openedx/core/djangoapps/content/learning_sequences/migrations/0014_remove_user_partition_group_duplicates.py

```diff
@@ -0,0 +1,33 @@
+# Generated by Django 2.2.24 on 2021-07-07 18:34
+
+from django.db import migrations
+from django.db.models import Count, Min
+
+
+class Migration(migrations.Migration):
+
+    def remove_user_partition_group_duplicates(apps, schema_editor):
+        UserPartitionGroup = apps.get_model('learning_sequences', 'UserPartitionGroup')
+
+        upg_duplicates = (
+            UserPartitionGroup.objects.values("partition_id", "group_id")
+            .annotate(first_entry_id=Min("id"), num_entries=Count("id"))
+            .filter(num_entries__gt=1)
+        )
+
+        # Delete all duplicates, excluding the original user partition group for the partition/group combo.
+        for duplicate in upg_duplicates:
+            (
+                UserPartitionGroup.objects
+                .filter(partition_id=duplicate["partition_id"], group_id=duplicate["group_id"])
+                .exclude(id=duplicate["first_entry_id"])
+                .delete()
+            )
```

Comment:
Does there need to be an update to move all the entries that pointed to the duplicate to point at the original?

Reply:
Excellent point, @ormsbee ! I'll look into that...

Reply:
OK - I added code to repoint existing entries and I tested it locally.

To test, I did the following steps:

With these existing course sections and user partition groups (containing duplicates):

```
mysql> select id,usage_key, title, course_context_id from learning_sequences_coursesection where id=1;
+----+------------------------------------------------------------------------------------+--------------+-------------------+
| id | usage_key                                                                          | title        | course_context_id |
+----+------------------------------------------------------------------------------------+--------------+-------------------+
|  1 | block-v1:edX+DemoX+Demo_Course+type@chapter+block@d8a6192ade314473a78242dfeedfbf5b | Introduction |                 1 |
+----+------------------------------------------------------------------------------------+--------------+-------------------+
1 row in set (0.00 sec)

mysql> select * from learning_sequences_userpartitiongroup;
+----+--------------+----------+
| id | partition_id | group_id |
+----+--------------+----------+
|  1 |            1 |        1 |
|  4 |            2 |        2 |
| 10 |            2 |        2 |
|  6 |            3 |        3 |
|  8 |            3 |        3 |
|  9 |            3 |        3 |
+----+--------------+----------+
6 rows in set (0.00 sec)
```
I added one of the duplicate user partition groups to the course section via the ManyToManyField:

```
>>> section.user_partition_groups.add(UserPartitionGroup.objects.filter(id=9)[0])
```
resulting in:
```
mysql> select * from learning_sequences_coursesection_user_partition_groups;
+----+------------------+-----------------------+
| id | coursesection_id | userpartitiongroup_id |
+----+------------------+-----------------------+
|  1 |                1 |                     9 |
+----+------------------+-----------------------+
1 row in set (0.00 sec)
```
Then I ran the data migration to test whether the course section was re-pointed to the single remaining non-duplicate user partition group.

And it was:
```
mysql> select * from learning_sequences_coursesection_user_partition_groups;
+----+------------------+-----------------------+
| id | coursesection_id | userpartitiongroup_id |
+----+------------------+-----------------------+
|  2 |                1 |                     6 |
+----+------------------+-----------------------+
1 row in set (0.00 sec)

mysql> select * from learning_sequences_userpartitiongroup;
+----+--------------+----------+
| id | partition_id | group_id |
+----+--------------+----------+
|  1 |            1 |        1 |
|  4 |            2 |        2 |
|  6 |            3 |        3 |
+----+--------------+----------+
3 rows in set (0.00 sec)
```

Reply:
Oh - I also tested it with multiple duplicates, using the following steps:

```
mysql> insert into learning_sequences_userpartitiongroup (group_id, partition_id) VALUES (3, 3);
Query OK, 1 row affected (0.00 sec)

mysql> insert into learning_sequences_userpartitiongroup (group_id, partition_id) VALUES (3, 3);
Query OK, 1 row affected (0.00 sec)

mysql> insert into learning_sequences_userpartitiongroup (group_id, partition_id) VALUES (3, 3);
Query OK, 1 row affected (0.01 sec)

mysql> insert into learning_sequences_userpartitiongroup (group_id, partition_id) VALUES (3, 3);
Query OK, 1 row affected (0.00 sec)

mysql> insert into learning_sequences_userpartitiongroup (group_id, partition_id) VALUES (3, 3);
Query OK, 1 row affected (0.00 sec)

mysql> select * from learning_sequences_userpartitiongroup;
+----+--------------+----------+
| id | partition_id | group_id |
+----+--------------+----------+
|  1 |            1 |        1 |
|  4 |            2 |        2 |
|  6 |            3 |        3 |
| 11 |            3 |        3 |
| 12 |            3 |        3 |
| 13 |            3 |        3 |
| 14 |            3 |        3 |
| 15 |            3 |        3 |
+----+--------------+----------+
8 rows in set (0.00 sec)

>>> section = CourseSection.objects.all()[0]
>>> section.user_partition_groups.add(UserPartitionGroup.objects.filter(id=12)[0])
>>> section.user_partition_groups.add(UserPartitionGroup.objects.filter(id=13)[0])
>>> section.user_partition_groups.add(UserPartitionGroup.objects.filter(id=14)[0])

mysql> select * from learning_sequences_coursesection_user_partition_groups;
+----+------------------+-----------------------+
| id | coursesection_id | userpartitiongroup_id |
+----+------------------+-----------------------+
|  2 |                1 |                     6 |
|  3 |                1 |                    12 |
|  4 |                1 |                    13 |
|  5 |                1 |                    14 |
+----+------------------+-----------------------+
4 rows in set (0.00 sec)

<Migration....>

mysql> select * from learning_sequences_coursesection_user_partition_groups;
+----+------------------+-----------------------+
| id | coursesection_id | userpartitiongroup_id |
+----+------------------+-----------------------+
|  2 |                1 |                     6 |
+----+------------------+-----------------------+
1 row in set (0.00 sec)

mysql> select * from learning_sequences_userpartitiongroup;
+----+--------------+----------+
| id | partition_id | group_id |
+----+--------------+----------+
|  1 |            1 |        1 |
|  4 |            2 |        2 |
|  6 |            3 |        3 |
+----+--------------+----------+
3 rows in set (0.00 sec)
```